core/rapid_test.go

```diff
@@ -0,0 +1,162 @@
+
+			nodes            = generateNodeAddresses(numNodes)
+			insertedBlocks   = make([]map[uint64][]byte, numNodes)
+			currentProposals = make([]uint64, numNodes)
+		)
+
+		// Initialize the block insertion map, used for lookups
+		for i := uint64(0); i < numNodes; i++ {
+			insertedBlocks[i] = make(map[uint64][]byte)
+		}
+
+		// commonTransportCallback is the common method modification
+		// required for Transport, for all nodes
+		commonTransportCallback := func(transport *mockTransport) {
+			transport.multicastFn = func(message *proto.Message) {
+				multicastFn(message)
+			}
+		}
+
+		// commonBackendCallback is the common method modification required
+		// for the Backend, for all nodes
+		commonBackendCallback := func(backend *mockBackend, nodeIndex int) {
+			// Make sure the quorum function requires all nodes
+			backend.quorumFn = func(_ uint64) uint64 {
+				return numNodes
+			}
+
+			// Make sure the node ID is properly relayed
+			backend.idFn = func() []byte {
+				return nodes[nodeIndex]
+			}
+
+			// Make sure the only proposer is picked using Round Robin
+			backend.isProposerFn = func(from []byte, height uint64, _ uint64) bool {
+				return bytes.Equal(from, nodes[height%numNodes])
+			}
+
+			// Make sure the proposal is valid if it matches what node 0 proposed
+			backend.isValidBlockFn = func(newProposal []byte) bool {
+				return bytes.Equal(newProposal, proposal)
+			}
+
+			// Make sure the proposal hash matches
+			backend.isValidProposalHashFn = func(p []byte, ph []byte) bool {
+				return bytes.Equal(p, proposal) && bytes.Equal(ph, proposalHash)
+			}
+
+			// Make sure the preprepare message is built correctly
+			backend.buildPrePrepareMessageFn = func(
+				proposal []byte,
+				certificate *proto.RoundChangeCertificate,
+				view *proto.View,
+			) *proto.Message {
+				return buildBasicPreprepareMessage(
+					proposal,
+					proposalHash,
+					certificate,
+					nodes[nodeIndex],
+					view)
+			}
+
+			// Make sure the prepare message is built correctly
+			backend.buildPrepareMessageFn = func(proposal []byte, view *proto.View) *proto.Message {
+				return buildBasicPrepareMessage(proposalHash, nodes[nodeIndex], view)
+			}
+
+			// Make sure the commit message is built correctly
+			backend.buildCommitMessageFn = func(proposal []byte, view *proto.View) *proto.Message {
+				return buildBasicCommitMessage(proposalHash, committedSeal, nodes[nodeIndex], view)
+			}
+
+			// Make sure the round change message is built correctly
+			backend.buildRoundChangeMessageFn = func(
+				proposal []byte,
+				certificate *proto.PreparedCertificate,
+				view *proto.View,
+			) *proto.Message {
+				return buildBasicRoundChangeMessage(proposal, certificate, view, nodes[nodeIndex])
+			}
+
+			// Make sure the inserted proposal is noted
+			backend.insertBlockFn = func(proposal []byte, _ []*messages.CommittedSeal) {
+				insertedBlocks[nodeIndex][currentProposals[nodeIndex]] = proposal
+				currentProposals[nodeIndex]++
```

Comment:
Is it going to be executed concurrently?

Reply:
Actually, yes, good catch 💯 

I've moved it into a thread safe structure to manage proposals:
[2ae44e7](https://github.com/0xPolygon/go-ibft/pull/27/commits/2ae44e79463befc2cfb25781feff11453aa55875)

Reply:
@JekaMas ping to resolve this convo if the discussion was answered